File: LAB-03/minbases.pl
% Helper predicate to check if a number is divisible by another
is_divisible(N, D) :- N mod D =:= 0.

% Recursive predicate to check if a number can be represented as a repdigit in the given base
is_base_valid(1, _) :- !.
is_base_valid(Num, Base) :- 
    is_divisible(Num - 1, Base),
    NextNum is (Num - 1) // Base,
    is_base_valid(NextNum, Base).

% Predicate to find the smallest base for a number
find_base(Num, Digit, Base) :- 
    search_base(Num, Digit + 1, Base).

% Helper predicate to search for the base
search_base(Num, Base, Base) :- 
    is_base_valid(Num, Base), !.
search_base(Num, Base, Result) :- 
    Base * Base > Num, !, 
    Result is Num - 1.
search_base(Num, Base, Result) :- 
    NextBase is Base + 1,
    search_base(Num, NextBase, Result).

% Predicate to solve for the minimal base of a given number
solve_min_base(1, 2) :- !.
solve_min_base(2, 3) :- !.
solve_min_base(N, MinBase) :-
    check_divisor(N, 1, N + 1, MinBase).

% Helper predicate to check divisors and find the minimal base
check_divisor(N, Divisor, CurrentMin, MinBase) :-
    Divisor * Divisor >= N, !,
    MinBase is CurrentMin.
check_divisor(N, Divisor, CurrentMin, MinBase) :-
    is_divisible(N, Divisor), !,
    Quotient is N // Divisor,
    find_base(Quotient, Divisor, PotentialBase),
    ( PotentialBase < CurrentMin 
    -> UpdatedMin is PotentialBase
    ;  UpdatedMin is CurrentMin),
    NextDivisor is Divisor + 1,
    check_divisor(N, NextDivisor, UpdatedMin, MinBase).
check_divisor(N, Divisor, CurrentMin, MinBase) :-
    NextDivisor is Divisor + 1,
    check_divisor(N, NextDivisor, CurrentMin, MinBase).

% Predicate to find the minimal bases for a list of numbers
find_min_bases([], []).
find_min_bases([Num|Nums], [Base|Bases]) :-
    solve_min_base(Num, Base),
    find_min_bases(Nums, Bases).

% Main entry point to find minimal bases for a list of numbers
minbases(Nums, Bases) :- find_min_bases(Nums, Bases).

File: LAB-02/minbases.ml
open Printf

let is_divisible n d = n mod d = 0

let rec is_base_valid num base = 
  match num with
  | 1 -> true
  | _ -> 
    if is_divisible (num - 1) base 
    && is_base_valid ((num - 1) / base) base 
    then true 
    else false

let find_base num digit =
  let rec search_base base =
    if is_base_valid num base 
    then base 
    else if base * base > num 
    then num - 1 
    else search_base (base + 1)
  in search_base (digit + 1)

let solve_min_base n =
  match n with
  | 1 -> 2
  | 2 -> 3
  | _ ->
    let rec check_divisor divisor current_min =
      let updated_min =
        if is_divisible n divisor then
          let potential_base = find_base (n / divisor) divisor in
          if potential_base < current_min && potential_base > divisor 
          then potential_base 
          else current_min
        else current_min
      in
      if divisor * divisor >= n 
      then updated_min 
      else check_divisor (divisor + 1) updated_min
    in check_divisor 1 (n + 1)

let find_min_bases num_list = List.map solve_min_base num_list

let () =
  (* Read input file name from command line argument *)
  let input_filename = Sys.argv.(1) in
  (* Open the input file and read all lines *)
  let lines = ref [] in
  let input_channel = open_in input_filename in
  try
    while true; do
      lines := input_line input_channel :: !lines
    done
  with End_of_file ->
    close_in input_channel;
    (* The first line contains the count of numbers, we ignore it *)
    let lines = List.rev !lines in
    match lines with
    | _ :: rest ->
      let numbers = List.filter_map (fun s -> 
        match int_of_string_opt s with
        | Some n -> Some n
        | None -> None) rest in
      (* Process each number to find the minimal base *)
      let results = find_min_bases numbers in
      (* Print results, each on a new line *)
      List.iter (printf "%d\n") results
    | _ -> ()